protocol-units/mempool/move-rocks/src/lib.rs

```diff
@@ -34,21 +34,36 @@ impl RocksdbMempool {
 		Ok(RocksdbMempool { db: Arc::new(RwLock::new(db)) })
 	}
 
-	pub fn construct_mempool_transaction_key(transaction: &MempoolTransaction) -> String {
-		// pad to 32 characters for slot_seconds
-		let slot_seconds_str = format!("{:032}", transaction.timestamp);
-
-		// pad to 32 characters for sequence number
-		let sequence_number_str = format!("{:032}", transaction.transaction.sequence_number);
 
+	pub fn construct_mempool_transaction_key(transaction: &MempoolTransaction) -> String {
+		// Pre-allocate a string with the required capacity
+		let mut key = String::with_capacity(32 + 1 + 32 + 1 + 32);
+	
+		// Function to pad an integer to 32 characters with leading zeros
+		fn pad_to_32_chars(num: u64) -> String {
+			let mut s = num.to_string();
+			while s.len() < 32 {
+				s.insert(0, '0');
+			}
+			s
+		}
```

Comment:
...but this is certainly worse than `format!("{num:032}")` because it repeatedly reallocates the temporary string and repeatedly inserts a character (a UTF-8 sequence, in fact) to the beginning with all of `O(N*2)` efficiency. Have you measured how this change improves performance?

Reply:
What string are you saying is reallocated? What is the $n$ in your $O(n^2)$?

Reply:
Oh, you're saying because the `s` is not allocated with the appropriate capacity.

Reply:
```rust
fn pad_to_32_chars(num: u64) -> String {

	let num_str = num.to_string();
    let num_len = num_str.len();

    // If the length is already 32 or more, return the string as is
    if num_len >= 32 {
        return num_str;
    }

    // Create a new String with a capacity of 32
    let mut padded_str = String::with_capacity(32);

    // Calculate the number of zeros needed
    let zeros_to_add = 32 - num_len;

    // Push the required number of zeros
    for _ in 0..zeros_to_add {
        padded_str.push('0');
    }

    // Push the original number string
    padded_str.push_str(&num_str);

    padded_str
}


#[test]
fn test_pad_to_32_chars_outperforms_format32() -> Result<(), anyhow::Error> {

// generate a vec of random u64s
let mut rng = rand::thread_rng();
let nums: Vec<u64> = (0..1000).map(|_| rng.gen()).collect();

// for each number time the two functions and store the results in a vec
let mut results = Vec::with_capacity(nums.len());
for num in nums.iter() {
	let start = std::time::Instant::now();
	let _ = pad_to_32_chars(*num);
	let pad_to_32_chars_time = start.elapsed();

	let start = std::time::Instant::now();
	let _ = format!("{:032}", num);
	let format32_time = start.elapsed();

	results.push((pad_to_32_chars_time, format32_time));
}

// check the mean time for each function
let pad_to_32_chars_mean = results.iter().map(|(a, _)| a).sum::<std::time::Duration>() / results.len() as u32;
let format32_mean = results.iter().map(|(_, b)| b).sum::<std::time::Duration>() / results.len() as u32;

// check that pad_to_32_chars is faster than format32
println!("pad_to_32_chars_mean: {:?}", pad_to_32_chars_mean);
println!("format32_mean: {:?}", format32_mean);
assert!(pad_to_32_chars_mean < format32_mean);

Ok(())

}

```

```
pad_to_32_chars_mean: 522ns
format32_mean: 245ns
```

Format is faster.